lsp/Main.hs

```diff
@@ -1,45 +1,151 @@
-            sendNotification SWindowShowMessage (ShowMessageParams MtError $ "Something went wrong!\n" <> T.pack (show err))
+      -- let params = ShowMessageRequestParams MtInfo "Turn on code lenses?"
+      --       (Just [MessageActionItem "Turn on", MessageActionItem "Don't"])
+      -- _ <- sendRequest SWindowShowMessageRequest params $ \case
+      --     Right (Just (MessageActionItem "Turn on")) -> do
+      --       let regOpts = CodeLensRegistrationOptions Nothing Nothing (Just False)
+
+      --       _ <- registerCapability STextDocumentCodeLens regOpts $ \_req responder -> do
+      --         let cmd = Command "Say hello" "lsp-hello-command" Nothing
+      --             rsp = List [CodeLens (mkRange 0 0 0 100) (Just cmd) Nothing]
+      --         responder (Right rsp)
+      --       pure ()
+      --     Right _ ->
+      --       sendNotification SWindowShowMessage (ShowMessageParams MtInfo "Not turning on code lenses")
+      --     Left err ->
+      --       sendNotification SWindowShowMessage (ShowMessageParams MtError $ "Something went wrong!\n" <> T.pack (show err))
+      elog "Hello world!"
       pure ()
   , requestHandler STextDocumentHover $ \req responder -> do
-      let RequestMessage _ _ _ (HoverParams _doc pos _workDone) = req
+      let RequestMessage _ _ _ (HoverParams doc pos _workDone) = req
           Position _l _c' = pos
           rsp = Hover ms (Just range)
           ms = HoverContents $ markedUpContent "lsp-demo-simple-server" "Hello world"
           range = Range pos pos
-      responder (Right $ Just rsp)
+      rsp <- lookupToken pos doc
+      -- responder (Right $ Just rsp)
+      responder (Right rsp)
+  -- , notificationHandler J.STextDocumentDidOpen $ \msg -> do
+  --   let doc  = msg ^. J.params . J.textDocument . J.uri
+  --       fileName =  J.uriToFilePath doc
+  --   liftIO $ debugM "reactor.handle" $ "Processing DidOpenTextDocument for: " ++ show fileName
+  --   sendDiagnostics (J.toNormalizedUri doc) (Just 0)
   ]
 
+-- | Analyze the file and send any diagnostics to the client in a
+-- "textDocument/publishDiagnostics" notification
+sendDiagnostics :: J.Range -> J.NormalizedUri -> Maybe Int -> LspM Config ()
+sendDiagnostics range fileUri version = do
+  let
+    diags = [J.Diagnostic
+              -- (J.Range (J.Position 0 1) (J.Position 0 5))
+              range
+              (Just J.DsWarning)  -- severity
+              Nothing  -- code
+              (Just "lsp-hello") -- source
+              "Example diagnostic message"
+              Nothing -- tags
+              (Just (J.List []))
+            ]
+  publishDiagnostics 100 fileUri version (partitionBySource diags)
+
+
+tshow :: Show a => a -> T.Text
+tshow = T.pack . show
+
+matchesPos' :: Int -> Int -> Token -> Bool
+matchesPos' line col (Token (AlexPn _ l c) len _) =
+  line == l -- && col == c
+  && col >= c && col < c + len
+
+elog :: (MonadIO m, Show a) => a -> m ()
+elog = liftIO . hPutStrLn stderr . tshow
```

Comment:
Is there standard logging infra in Haskell?

Reply:
Yes, I just wanted to get some output as quickly as possible.
There are some examples in `Reactor.hs`.